File: backend/app/schemas/analytics.py
"""
Pydantic schemas for Analytics API.
"""
from typing import Dict, List
from pydantic import BaseModel


class ApplicationsByStatus(BaseModel):
    """Count of applications by status."""

    Applied: int = 0
    Interviewing: int = 0
    Offer: int = 0
    Rejected: int = 0


class ApplicationsOverTime(BaseModel):
    """Applications count over time."""

    date: str  # Format: YYYY-MM
    count: int


class AverageTimePerStage(BaseModel):
    """Average days spent in each stage."""

    Applied: float = 0.0
    Interviewing: float = 0.0


class AnalyticsResponse(BaseModel):
    """Analytics dashboard response."""

    total_applications: int
    by_status: ApplicationsByStatus
    applications_over_time: List[ApplicationsOverTime]
    average_time_per_stage: AverageTimePerStage
    success_rate: float  # Offers / Total applications

File: backend/src/schemas/analytics.ts
/**
 * Zod schemas for analytics validation
 */
import { z } from 'zod';

/**
 * Count of applications by status
 */
export const applicationsByStatusSchema = z.object({
  Applied: z.number().int().default(0),
  Interviewing: z.number().int().default(0),
  Offer: z.number().int().default(0),
  Rejected: z.number().int().default(0),
});

/**
 * Applications count over time
 */
export const applicationsOverTimeSchema = z.object({
  date: z.string(), // Format: YYYY-MM
  count: z.number().int(),
});

/**
 * Average days spent in each stage
 */
export const averageTimePerStageSchema = z.object({
  Applied: z.number().default(0.0),
  Interviewing: z.number().default(0.0),
});

/**
 * Analytics dashboard response
 */
export const analyticsResponseSchema = z.object({
  totalApplications: z.number().int(),
  byStatus: applicationsByStatusSchema,
  applicationsOverTime: z.array(applicationsOverTimeSchema),
  averageTimePerStage: averageTimePerStageSchema,
  successRate: z.number(), // Offers / Total applications
});

/**
 * TypeScript types from Zod schemas (for validation only)
 * Use types from @jackdo69/job-tracker-shared-types for application logic
 */
export type AnalyticsResponse = z.infer<typeof analyticsResponseSchema>;
